src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,284 @@
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (generationsLeft == -1)
+        {
+            TryToTriggerEvent(totalTimePassed);
+        }
+        else if (generationsLeft == 0)
+        {
+            FinishEvent();
+        }
+    }
+
+    private bool IsSurfacePatch(Patch patch)
+    {
+        return patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave;
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (IsSurfacePatch(patch))
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (generationsToTrigger == -1)
+        {
+            if (!AreConditionsMet())
+                return;
+
+            generationsToTrigger = Constants.GLOBAL_GLACIATION_HEADS_UP_DURATION;
+            return;
+        }
+
+        switch (generationsToTrigger)
+        {
+            case > 0:
+                LogHeadUpEventWarning();
+                generationsToTrigger -= 1;
+                break;
+            case 0:
+            {
+                eventDuration = random.Next(Constants.GLOBAL_GLACIATION_MIN_DURATION,
+                    Constants.GLOBAL_GLACIATION_MAX_DURATION);
+                generationsLeft = eventDuration;
+
+                foreach (var (index, patch) in targetWorld.Map.Patches)
+                {
+                    if (IsSurfacePatch(patch))
+                    {
+                        ChangePatchProperties(index, patch, totalTimePassed);
+                    }
+                }
+
+                LogBeginningOfGlaciation();
+                break;
+            }
+        }
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (!IsSurfacePatch(patch))
+                continue;
+
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var oxygenLevel);
+            if (oxygenLevel.Ambient >= Constants.GLOBAL_GLACIATION_OXYGEN_THRESHOLD)
+                patchesExceedingOxygenLevel += 1;
+
+            numberOfSurfacePatches += 1;
+        }
+
+        // Just prevent dividing by zero, but that shouldn't be possible anyway
+        numberOfSurfacePatches = numberOfSurfacePatches == 0 ? 1 : numberOfSurfacePatches;
+
+        return (float)patchesExceedingOxygenLevel / numberOfSurfacePatches >=
+            Constants.GLOBAL_GLACIATION_PATCHES_THRESHOLD
+            && random.NextFloat() <= Constants.GLOBAL_GLACIATION_CHANCE;
+    }
+
+    private void ChangePatchProperties(int index, Patch patch, double totalTimePassed)
+    {
+        modifiedPatchesIds.Add(index);
```

Comment:
Index is not the ID. So this code is incorrect: ` ChangePatchProperties(index, patch, totalTimePassed);`

You should actually use the patch ID.